nicegui/element_filter.py

```diff
@@ -0,0 +1,223 @@
+
+from typing import Generic, Iterator, List, Optional, Type, TypeVar, Union, overload
+
+from typing_extensions import Self
+
+from .context import context
+from .element import Element
+from .elements.mixins.content_element import ContentElement
+from .elements.mixins.source_element import SourceElement
+from .elements.mixins.text_element import TextElement
+
+T = TypeVar('T', bound=Element)
+
+
+class ElementFilter(Generic[T]):
+    DEFAULT_LOCAL_SCOPE = False
+
+    @overload
+    def __init__(self: ElementFilter[Element], *,
+                 marker: Union[str, List[str], None] = None,
+                 content: Union[str, List[str], None] = None,
+                 local_scope: bool = DEFAULT_LOCAL_SCOPE,
+                 ) -> None:
+        ...
+
+    @overload
+    def __init__(self, *,
+                 kind: Type[T],
+                 marker: Union[str, List[str], None] = None,
+                 content: Union[str, List[str], None] = None,
+                 local_scope: bool = DEFAULT_LOCAL_SCOPE,
+                 ) -> None:
+        ...
+
+    def __init__(self, *,
+                 kind: Optional[Type[T]] = None,
+                 marker: Union[str, List[str], None] = None,
+                 content: Union[str, List[str], None] = None,
+                 local_scope: bool = DEFAULT_LOCAL_SCOPE,
+                 ) -> None:
+        """ElementFilter
+
+        Sometimes it is handy to search the Python element tree of the current page.
+        ``ElementFilter()`` allows powerful filtering by kind of elements, markers and content.
+        It also provides a fluent interface to apply more filters like excluding elements or filtering for elements within a specific parent.
+        The filter can be used as an iterator to iterate over the found elements and is always applied while iterating and not when being instantiated.
+
+        And element is yielded if it matches all of the following conditions:
+
+        - The element is of the specified kind (if specified).
+        - The element is not of any of the excluded kinds.
+        - The element has all of the specified markers.
+        - The element has none of the excluded markers.
+        - The element contains all of the specified content.
+        - The element contains none of the excluded content.
+
+        - The ancestors of the element include all of the specified instances defined via ``within``.
+        - The ancestors of the element do not include any of the specified instances defined via ``not_within``.
+        - The ancestors of the element include all of the specified kinds defined via ``within``.
+        - The ancestors of the element do not include any of the specified kinds defined via ``not_within``.
+        - The ancestors of the element include at least one element with all of the specified markers defined via ``within``.
+        - The ancestors of the element do not include any element with all of the specified markers defined via ``not_within``.
```

Comment:
```suggestion
        - The ancestors of the element include all of the specified instances defined via ``within``.
        - The ancestors of the element do not include any of the specified instances defined via ``not_within``.
        - The ancestors of the element include all of the specified kinds defined via ``within``.
        - The ancestors of the element do not include any of the specified kinds defined via ``not_within``.
        - The ancestors of the element include all of the specified markers defined via ``within``.
        - The ancestors of the element do not include any of the specified markers defined via ``not_within``.
```
@rodja Now that I have fixed the rule for ancestor kinds, I wonder if we should also change the last two rules for symmetry. Currently they enforce single elements to include or exclude all markers at once. But I feel it is easier to understand if the required markers apply to the whole ancestor stack.

Reply:
Yes, that is much more consistent!

Reply:
Ok, I'll change the implementation and pytests accordingly.